### PR TITLE
chore: datafusion 34, arrow & parquet 49

### LINKED ISSUE #1805: Delta Stats for binary columns are not truncated
# Environment

**Delta-rs version**: master

**Binding**: rust

**Environment**: local test

***
# Bug

**What happened**:

When writing a file with large binary columns, the delta log json for the commit is very large due to a large statistics object.

**What you expected to happen**:

These are expected to receive truncated statistics due to a PR to arrow-rs (https://github.com/apache/arrow-rs/pull/4389).

**How to reproduce it**:

A test case (to put in `stats.rs`):

```rust


    #[tokio::test]
    async fn test_delta_stats_truncation() -> Result<(), crate::DeltaTableError> {
        let temp_dir = tempfile::tempdir().unwrap();
        let table_path = temp_dir.path().to_owned();

        let schema_fields = vec![
            crate::schema::SchemaField::new(
                "long_string".to_owned(),
                crate::SchemaDataType::primitive("string".to_owned()),
                false,
                Default::default(),
            ),
            crate::schema::SchemaField::new(
                "long_binary".to_owned(),
                crate::SchemaDataType::primitive("binary".to_owned()),
                false,
                Default::default(),
            ),
        ];

        let table = crate::operations::create::CreateBuilder::new()
            .with_table_name("temp")
            .with_location(table_path.to_str().unwrap())
            .with_columns(schema_fields.clone())
            .await?;
        let mut writer = RecordBatchWriter::for_table(&table).unwrap();
        writer = writer.with_writer_properties(
            WriterProperties::builder()
                .set_compression(Compression::SNAPPY)
                .set_max_row_group_size(128)
                .build(),
        );

        let fields = arrow::datatypes::Fields::from(
            schema_fields
                .into_iter()
                .map(|f| arrow::datatypes::Field::try_from(&f).unwrap())
                .collect::<Vec<_>>(),
        );

        let long_field_len =
            10 * parquet::file::properties::DEFAULT_COLUMN_INDEX_TRUNCATE_LENGTH.unwrap();

        const ROW_COUNT: usize = 10;

        let mut string_builder = arrow::array::StringBuilder::new();
        let mut binary_builder = arrow::array::BinaryBuilder::new();

        for i in 0..ROW_COUNT {
            let long_string = std::iter::repeat(i.to_string())
                .take(long_field_len)
                .collect::<Vec<_>>()
                .join("");
            string_builder.append_value(&long_string);

            let long_binary = std::iter::repeat(i as u8)
                .take(long_field_len)
                .collect::<Vec<_>>();
            binary_builder.append_value(&long_binary);
        }

        let arrays: Vec<Arc<dyn arrow::array::Array>> = vec![
            Arc::new(string_builder.finish()),
            Arc::new(binary_builder.finish()),
        ];

        let file_contents: arrow::record_batch::RecordBatch =
            StructArray::new(fields, arrays, None).into();

        writer.write(file_contents).await?;

        let mut actions = writer.flush().await?;

        assert!(actions.len() == 1);

        let action = actions.remove(0);

        let stats = action.get_stats()?.expect("stats");

        match stats.min_values.get("long_string").unwrap() {
            ColumnValueStat::Value(serde_json::Value::String(s)) => {
                assert_eq!(
                    s.len(),
                    parquet::file::properties::DEFAULT_COLUMN_INDEX_TRUNCATE_LENGTH.unwrap()
                );
            }
            x => panic!("invalid stats format: {x:?}"),
        }

        match stats.min_values.get("long_binary").unwrap() {
            ColumnValueStat::Value(serde_json::Value::String(s)) => {
                assert_eq!(
                    s.len(),
                    parquet::file::properties::DEFAULT_COLUMN_INDEX_TRUNCATE_LENGTH.unwrap()
                );
            }
            x => panic!("invalid stats format: {x:?}"),
        }

        Ok(())
    }

```

**More details**:

I think this is due to the underlying parquet write truncating the values it uses for the column index, but not for the column metadata statistics. I'm going to open a companion issue to arrow-rs to track that. (EDIT: https://github.com/apache/arrow-rs/issues/5037)

### PR DESCRIPTION
**Pull request by emcake**
# Description
Upgrades DataFusion and Arrow to v34 and v49 respectively. As a casualty we also need to upgrade object_store to v0.8.

# Related Issue(s)
Arrow 49 contains a fix to allow for truncated statistics on binary columns (https://github.com/apache/arrow-rs/issues/5037) that I'd like to employ to fix #1805 .

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -19,24 +15,26 @@ debug = "line-tables-only"
...
+arrow-array = { version = "49.0.0" }
+arrow-buffer = { version = "49.0.0" }
+arrow-cast = { version = "49.0.0" }
+arrow-ord = { version = "49.0.0" }
+arrow-row = { version = "49.0.0" }
+arrow-schema = { version = "49.0.0" }
+arrow-select = { version = "49.0.0" }
+parquet = { version = "49.0.0" }
+
+object_store = "0.8"
```
Comment (emcake) [inline Cargo.toml:29]: I moved `object_store` to a workspace dependency as it's referenced in two separate crates - easier to keep the dependency in sync this way.

**Comment by roeap** [comment]
@emcake - thanks for taking care of this!

Seems there is one more clippy error to address :).

**Comment by emcake** [comment]
I think this is now complete and up to date - there still appears to be some tests failing but I'm struggling to work out if these are flaky or real.

**Comment by ion-elgreco** [comment]
@emcake the restore by datetime and a read test are super flaky..

**Comment by ion-elgreco** [comment]
@emcake maybe object store 0.8 is causing the read failures, can you revert to object store 0.7?

**Comment by roeap** [comment]
Object store is coupled to datafusion. Had a first glance, and it seems like fixable problem, but may require some work.

**Comment by emcake** [comment]
Looks like @roeap beat me to it!

**Pull request closed without merging**